#517151 [C#テキスト]2章演習問題

sugisaka-being opened the pull request:
2章演習問題の解答を作成しました。

sugisaka-being committed:
#517151 [C#テキスト]2章演習問題

2章演習問題の解答を作成しました。

m-hotta-being commented on Chapter2/Chapter2-1-2/Program2-1-2.cs:
```diff
@@ -0,0 +1,16 @@
+        　　テキストに記載した「距離換算プログラム」のコードを参考に、インチからメートルへの変換表を1インチ刻みでコンソールに出力するプログラムを書いてください。
+        　　このときのインチの範囲は、1インチから10インチまでとしてください。1インチは0.0254メートルです。
+        */
+        static void Main(string[] args) {
+
+            for (int inch = 1; inch <= 10; inch++) {
+                Console.WriteLine($"{inch}インチは{InchConverter.ToMeter(inch)}メートルです。");
+            }
+        }
```
【FYI: 情報共有。アクション不要です。】
for, foreach, while を使わずに書いてみることに挑戦してみても面白いですよ。

teshima-being commented on Chapter2/Chapter2-1-1/Song.cs:
```diff
@@ -0,0 +1,35 @@
+        /// </summary>
+        /// <param name="vTitle">歌のタイトル</param>
+        /// <param name="vArtistname">アーティスト名</param>
+        /// <param name="vLength">演奏時間（単位は秒）</param>
```
演奏時間の単位についての説明があるのは親切でよいと思います！

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,39 @@
+        static void Main(string[] args) {
+
+            // 3.
+            Song[] wSongs = new Song[] {
+            new Song("000", "aaa", 60),
+            new Song("111", "bbb", 90),
+            new Song("222", "ccc", 120),
```
※追加課題
今回の配列とは別に、`List<Song> wSongsList`を用意し、そこに3つほどSongインスタンスを格納してみてください。
また、配列とListの違いについてもこの機会に調べてみてください。（すでに確認済ならOKです。）

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,39 @@
+
+            // 4.
+            foreach (var song in wSongs) {
+                Console.WriteLine($"歌のタイトル：{song.Title}、アーティスト名：{song.ArtistName}、演奏時間{song.Length/60}:{song.Length%60:00}");
```
※追加課題
演奏時間の出力について、`{song.Length/60}:{song.Length%60:00}`にしてくれていますが、TimeSpan構造体を使って書き直してみてください。
参考：[TimeSpan構造体](https://learn.microsoft.com/ja-jp/dotnet/api/system.timespan?view=net-7.0)
　　　[【C#】秒数をhh:mm:ss変換する](https://qiita.com/Nossa/items/70487b765ec9332e0db0)

teshima-being commented on Chapter2/Chapter2-1-1/Program2-1-1.cs:
```diff
@@ -0,0 +1,39 @@
+            };
+
+            // 4.
+            foreach (var song in wSongs) {
```
`song`はローカル変数なので接頭語の`w`が必要ですね。

teshima-being commented on Chapter2/Chapter2-1-2/InchConverter.cs:
```diff
@@ -0,0 +1,17 @@
+    /// インチ変換クラス
+    /// </summary>
+    internal class InchConverter {
+        private const double ratio = 0.0254;
```
`ratio`は定数なので、[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#:~:text=%E4%BE%8B%EF%BC%9Astring%20wMessage-,%E5%AE%9A%E6%95%B0,-C_)に則り接頭語の`C_`を付けましょう。

定数として変換率を定義しているのは非常に良いと思います。
今回のケースは今後数値が変わることはないかと思いますが、今後変わる可能性がある数値を複数個所で直接記述していると、数値に変更があった場合にそのすべてを書き換える必要が発生し、変更漏れによるバグなどに繋がります。
そういった数値を定数として定義しておくことで、数値に変更があった場合でも修正箇所は定数を定義している1箇所で済みます。

teshima-being commented on Chapter2/Chapter2-1-2/Program2-1-2.cs:
```diff
@@ -0,0 +1,16 @@
+        static void Main(string[] args) {
+
+            for (int inch = 1; inch <= 10; inch++) {
+                Console.WriteLine($"{inch}インチは{InchConverter.ToMeter(inch)}メートルです。");
```
`inch`はローカル変数なので接頭語`w`が必要ですね。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,54 @@
+    /// 売上集計クラス
+    /// </summary>
+    internal class SalesCounter {
+        private IEnumerable<Sale> _sales;
```
salesはprivateなフィールド変数なので[コーディング規約](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#%E6%8E%A5%E9%A0%AD%E8%BE%9E)に則り、接頭語Fを付けましょう。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,54 @@
+                    ShopName = wItems[0],
+                    ProductCategory = wItems[1],
+                    Amount = int.Parse(wItems[2])
+                };
```
※追加課題
現在は初期化子{};でSaleインスタンスの各プロパティを設定していますが、Saleクラスにコンストラクタを実装し、コンストラクタを使って書き直してみてください。
コンストラクタを実装することで、Saleクラスのインスタンスを作成するときに、必要なプロパティを強制的に設定させることができるというメリットがあります。

現在のままだと下記のように、例えば店名がないSaleインスタンスを作成することが出来ますが、コンストラクタを適切に実装することで、そういったことを防ぐことが出来ます。

                var wSale = new Sale {
                    ProductCategory = wItems[1],
                    Amount = int.Parse(wItems[2])
                };

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,54 @@
+            var wSales = new List<Sale>();
+            var wLines = File.ReadAllLines(vFilePath);
+            foreach (var line in wLines) {
+                var wItems = line.Split(',');
```
`line`はローカル変数なので接頭語の`w`が必要です。

teshima-being commented on Chapter2/Chapter2-1-3/Program2-1-3.cs:
```diff
@@ -0,0 +1,16 @@
+        static void Main(string[] args) {
+            var wSales = new SalesCounter("C:..\\..\\sales.csv");
+            var wAmountPerCategory = wSales.GetPerProductSales();
+            foreach (var obj in wAmountPerCategory) {
+                Console.WriteLine($"商品カテゴリ：{obj.Key}、売上高：{obj.Value}");
```
`obj`はローカル変数なので接頭語の`w`が必要です。
また、テキストの例が`obj`になっているのかもしれませんが、この機会に変数名としてより良いものがないかも考えてみてください。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,54 @@
+        /// </summary>
+        /// <returns>商品カテゴリ別の売上高</returns>
+        public IDictionary<string, int> GetPerProductSales() {
+            var wDict = new Dictionary<string, int>();
```
`wDict`についてもより良い変数名がないか検討してみてください。
`wDict`だとなんの情報を持っているのか名前からは判断がつかないです。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -0,0 +1,54 @@
+        public IDictionary<string, int> GetPerProductSales() {
+            var wDict = new Dictionary<string, int>();
+            foreach (var sale in _sales) {
+                if (wDict.ContainsKey(sale.ProductCategory))
+                    wDict[sale.ProductCategory] += sale.Amount;
+                else
+                    wDict[sale.ProductCategory] = sale.Amount;
+            }
```
※参考情報
キーの存在確認の方法として、`ContainsKey`のかわりに[TryGetValueメソッド](https://learn.microsoft.com/ja-jp/dotnet/api/system.collections.generic.idictionary-2.trygetvalue?view=net-8.0)を使うこともできます。
キーの存在確認後に値を取得したいという場合、`TryGetValue`を使えば両方を1度の処理で行うことが出来ます。

teshima-being commented on Chapter2/Chapter2-1-3/Program2-1-3.cs:
```diff
@@ -0,0 +1,16 @@
+        　　テキストに記載した「売上集計プログラム」のプログラムを変更し、商品カテゴリ別の売上高を求めるプログラムを作成してください。
+        */
+        static void Main(string[] args) {
+            var wSales = new SalesCounter("C:..\\..\\sales.csv");
```
相対パスでファイルのパスを指定する場合、`C:`の部分は不要です。Cドライブ以外のドライブで実行する場合にパスが機能しない可能性があります。

※以下は参考情報
エスケープシーケンスについて調べてもらったかもしれませんが、パスを記述する際に、毎回`\\`を2回記述していると見た目が少し複雑になりがちです。
そういう場合は@を利用して、`new SalesCounter(@"..\..\sales.csv")`と記述することもできます。この方が見た目は分かりやすくなります。
また、時間があるときに「逐語的文字列リテラル」で調べてみてください。

teshima-being reviewed (commented):
とても見やすいコードでした！
いくつか細かい指摘と追加の課題をコメントしましたので、確認お願いいたします。

sugisaka-being committed:
#517151 [C#テキスト]2章演習問題

追加課題に取り組みました。
→Chapter2-1-1/Program2-1-1.csにおいて、List<Song> wSongsListを用意し、そこに3つのSongインスタンスを格納しました。
→Chapter2-1-1のProgram2-1-1.csにおいて、TimeSpan構造体を利用しました。
→Chapter2-1-3において、Saleクラスにコンストラクタを実装してからインスタンスを作成しました。
コーディング規約に則り接頭語を付けました。
→Chapter2-1-1のProgram2-1-1.csにおける「song」
→Chapter2-1-2のInchConverter.csにおける「ratio」
→Chapter2-1-2のProgram2-1-2.csにおける「inch」
→Chapter2-1-3のSalesCounter.csにおける「sale」
→Chapter2-1-3のSalesCounter.csにおける「line」
→Chapter2-1-3のProgram2-1-3.csにおける「obj」。また、わかりやすいものへと変数名を変更しました「→wSale」
どんな情報を持っているかをすぐに判断できるように変数名を変更しました。
→Chapter2-1-3のSalesCounter.csにおいて、「wDict→ｗSales」
相対パスでファイルのパスを指定する際に用いていた不要な「C:」を削除しました。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -24,13 +24,9 @@ public SalesCounter(string vFilePath) {
-                    ProductCategory = wItems[1],
-                    Amount = int.Parse(wItems[2])
-                };
+            foreach (var wLine in wLines) {
+                var wItems = wLine.Split(',');
+                var wSale = new Sale(wItems[0], wItems[2], int.Parse(wItems[3]));
```
Saleクラスのコンストラクタを実装してくれたようですね！
1点だけすごく惜しいところがあります。
一度プログラムを実行してみて動作を確認したのち、修正をお願いします。

sugisaka-being replied:
ありがとうございます。
`var wSale = new Sale(wItems[0], wItems[2], int.Parse(wItems[3]));`を
`var wSale = new Sale(wItems[0], wItems[1], int.Parse(wItems[2]));`に修正いたしました。

teshima-being commented on Chapter2/Chapter2-1-3/SalesCounter.cs:
```diff
@@ -41,14 +37,14 @@ private static IEnumerable<Sale> ReadSales(string vFilePath) {
-            foreach (var sale in _sales) {
-                if (wDict.ContainsKey(sale.ProductCategory))
-                    wDict[sale.ProductCategory] += sale.Amount;
+            var wSales = new Dictionary<string, int>();
```
※参考（修正不要）
wDictから変更したことで売上を示すものであることが分かるようになりましたね。
今回は商品ごとの売上をDictionaryに格納しているので、SalesPerProductとかだとより情報量が多いかなと思います。

sugisaka-being replied:
ありがとうございます。
その変数が何を示しているのかをより分かりやすくため、変数名を`wSalesPerProduct`に修正いたしました。

teshima-being commented on Chapter2/Chapter2-1-3/Program2-1-3.cs:
```diff
@@ -6,10 +6,10 @@ internal class Program {
         */
         static void Main(string[] args) {
-            var wSales = new SalesCounter("C:..\\..\\sales.csv");
+            var wSales = new SalesCounter(@"..\..\sales.csv");
```
※参考情報
本来はプログラムの中にファイルパスを直接記述するのは得策とは言えません。対象ファイルを移動するとプログラムが動かなくなってしまう場合があります。
そういう場合はプログラム実行時にユーザーがファイルパスを都度指定できるとより使い勝手がよくなります。
テキストの課題の枠を超えて、`Console.ReadLine()`やコマンドライン引数を積極的に利用してみてください。

sugisaka-being replied:
`Console.ReadLine()`を用いてファイルパスを指定する方法を試し、ファイルパスを直接記述した際と同様に動作することを確認いたしました。
コマンドライン引数を利用することでファイルパスを都度指定することが可能になるのですね。
非常に勉強になります。ありがとうございます。

sugisaka-being replied:
Console.ReadLine()を用いてファイルパスを指定する方法を試し、ファイルパスを直接記述した際と同様に動作することを確認いたしました。

teshima-being reviewed (commented):
あと少しだけ修正をお願いします！

sugisaka-being committed:
#517151 [C#テキスト]2章演習問題

変数名を変更しました。
Saleクラスのコンストラクタの実装部分を修正いたしました。

teshima-being commented:
レビューOKです。2章お疲れ様でした。
